air/src/execution_step/value_types/canon_stream_map.rs

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright 2023 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use super::stream_map::VALUE_FIELD;
+use super::CanonStream;
+use super::TracePosOperate;
+use super::ValueAggregate;
+use crate::execution_step::execution_context::stream_map_key::StreamMapKey;
+use crate::execution_step::ExecutionResult;
+use crate::ExecutionError;
+use crate::JValue;
+use crate::StreamMapKeyError::UnsupportedKVPairObjectOrMapKeyType;
+use crate::UncatchableError;
+
+use air_interpreter_cid::CID;
+use air_interpreter_data::CanonResultCidAggregate;
+use polyplets::SecurityTetraplet;
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::rc::Rc;
+
+/// Canon stream map is a read-only struct that mimics conventional map.
+/// The contents of a map are fixed at a certain node.
+#[derive(Debug, Clone)]
+pub struct CanonStreamMap<'key> {
+    // values vec contains all KVpair objects {"key": key, "value": value}.
+    // There might be multiple KVPars with the same key.
+    values: Vec<ValueAggregate>,
+    // Index access leverages the map that does key to CanonStream mapping.
+    map: HashMap<StreamMapKey<'key>, CanonStream>,
+    // ap arg processing leverages this tetraplet
+    tetraplet: Rc<SecurityTetraplet>,
+}
+
+impl<'key> CanonStreamMap<'key> {
+    // The argument's tetraplet is used to produce canon streams for keys so
+    // that the produced canon streams share tetraplets with the original canon stream
+    // rendered by canon instruction.
+    pub(crate) fn from_canon_stream(canon_stream: CanonStream) -> ExecutionResult<CanonStreamMap<'key>> {
+        let mut map: HashMap<StreamMapKey<'_>, CanonStream> = HashMap::new();
+        let tetraplet = canon_stream.tetraplet.clone();
+        for kvpair_obj in canon_stream.iter() {
```

Comment:
It's better to rewrite this for with a monadic flow.

Reply:
Do you have something particular in mind? For example fold() used here makes this part unreadable.

Reply:
The result here is `ExecutionResult<HashMap<_, Vec<_>>>`.  Itertools has methods that group items, producing `HashMap<_, Vec<_>>`, but they do not handle results.